File: log.go
package main

import (
	"encoding/json"
	"os"
	"path/filepath"
	"sort"
)

type uploadLogData struct {
	Uploaded []string          `json:"uploaded"`
	Albums   map[string]string `json:"albums"`
}

func LoadUploadLog(exportDir string) (map[string]bool, map[string]string, error) {
	logPath := filepath.Join(exportDir, UploadLog)
	data, err := os.ReadFile(logPath)
	if err != nil {
		if os.IsNotExist(err) {
			return make(map[string]bool), make(map[string]string), nil
		}
		return nil, nil, err
	}

	var logData uploadLogData
	if err := json.Unmarshal(data, &logData); err != nil {
		return nil, nil, err
	}

	uploaded := make(map[string]bool, len(logData.Uploaded))
	for _, path := range logData.Uploaded {
		uploaded[path] = true
	}

	albums := logData.Albums
	if albums == nil {
		albums = make(map[string]string)
	}
	return uploaded, albums, nil
}

func SaveUploadLog(exportDir string, uploaded map[string]bool, albums map[string]string) error {
	logPath := filepath.Join(exportDir, UploadLog)

	paths := make([]string, 0, len(uploaded))
	for p := range uploaded {
		paths = append(paths, p)
	}
	sort.Strings(paths)

	logData := uploadLogData{
		Uploaded: paths,
		Albums:   albums,
	}

	data, err := json.MarshalIndent(logData, "", "  ")
	if err != nil {
		return err
	}
	data = append(data, '\n')
	return os.WriteFile(logPath, data, 0644)
}

File: photo_export/log.py
import json
from pathlib import Path

from config import UPLOAD_LOG


def load_upload_log(export_dir: Path) -> tuple[set, dict]:
    """Загружает список ранее загруженных файлов и кеш альбомов."""
    log_path = export_dir / UPLOAD_LOG
    if log_path.exists():
        with open(log_path) as f:
            data = json.load(f)
            return set(data.get("uploaded", [])), data.get("albums", {})
    return set(), {}


def save_upload_log(export_dir: Path, uploaded: set, albums: dict):
    """Сохраняет список загруженных файлов и кеш альбомов."""
    log_path = export_dir / UPLOAD_LOG
    with open(log_path, "w") as f:
        json.dump({"uploaded": sorted(uploaded), "albums": albums},
                  f, indent=2, ensure_ascii=False)
